### PR TITLE
Add serialization constructor for AssertionFailedException

### PR DESCRIPTION
**Pull request by florianwittmann**
Currently the  AssertionFailedException is missing a serializable constructor. Therefore it throws an error when deserializing the exception.

See: http://stackoverflow.com/questions/94488/what-is-the-correct-way-to-make-a-custom-net-exception-serializable

Addition to implement #214 properly

### CODE REVIEW

**Comment by dennisdoomen** [comment]
Thanks. Unfortunately, it contains some StyleCop warnings (that are treated as errors these days). Could you please fix those?

**Comment by florianwittmann** [comment]
Sure, working on it.

**Comment by florianwittmann** [comment]
Okay the unit test is failing (Like in manual test)
Test method 

>Type 'FluentAssertions.Execution.AssertionFailedException' in Assembly 'FluentAssertions, Version=3.4.0.4, Culture=neutral, PublicKeyToken=33f2691a05b67b6a' is not marked as serializable

Actually it is:
>  #if NET40 || NET45
     [Serializable]
 #endif

Can't get the error here?
Commenting out the conditionals in AssertionFailedException, will make the test green for Net40 and Net45, but will have the other platforms to throw compile errors.

**Comment by dennisdoomen** [comment]
Shouldn't you wrap that unit test in `#if NET40 || NET45`

File: FluentAssertions.Shared.Specs/EventAssertionSpecs.cs [hunk 1]
```diff
@@ -578,7 +583,6 @@ public void When_a_monitored_class_in_not_referenced_anymore_it_should_be_garbag
         }
 
 #if NET40 || NET45
```
Comment (florianwittmann) [inline FluentAssertions.Shared.Specs/EventAssertionSpecs.cs]: Should not this 
> #if NET40 || NET45
> [...]
> #endif

 block already span both unit tests?

**Comment by florianwittmann** [comment]
But yes, sure it's for NET40 and NET45 only.

**Comment by dennisdoomen** [comment]
I found the problem. The .NET 4.0 assembly doesn't define the NET40 symbol, so the `[Serializable]` attribute is not included.

**Comment by florianwittmann** [comment]
Okay, passing now, but a lots of commits, I would prepare a clean fresh pull request.

**Pull request closed without merging**